### ISSUE TITLE
Supervisord fails because of missing dependancy

### ISSUE THREAD

**Issue by stevenazari** [puphpet/puphpet#1492]
Supervisord is failing because it's missing a dependancy and thus vagrant breaks.

==> default: Notice: /Stage[main]/Supervisord::Install/Package[supervisor]/ensure: created
==> default: Notice: /Stage[main]/Supervisord::Config/File[/var/log/supervisor]/ensure: created
==> default: Notice: /Stage[main]/Supervisord::Config/File[/etc/default/supervisor]/ensure: created
==> default: Notice: /Stage[main]/Supervisord::Config/File[/etc/init.d/supervisord]/ensure: created
==> default: Notice: /Stage[main]/Supervisord::Config/File[/etc/supervisor.d]/ensure: created
==> default: Notice: /Stage[main]/Supervisord::Config/Concat[/etc/supervisord.conf]/File[/var/lib/puppet/concat/_etc_supervisord.conf]/ensure: created
==> default: Info: /Stage[main]/Supervisord::Config/Concat[/etc/supervisord.conf]/File[/var/lib/puppet/concat/_etc_supervisord.conf]: Scheduling refresh of Exec[concat_/etc/supervisord.conf]
==> default: Notice: /Stage[main]/Supervisord::Config/Concat[/etc/supervisord.conf]/File[/var/lib/puppet/concat/_etc_supervisord.conf/fragments]/ensure: created
==> default: Info: /Stage[main]/Supervisord::Config/Concat[/etc/supervisord.conf]/File[/var/lib/puppet/concat/_etc_supervisord.conf/fragments]: Scheduling refresh of Exec[concat_/etc/supervisord.conf]
==> default: Notice: /Stage[main]/Supervisord::Config/Concat::Fragment[supervisord_main]/File[/var/lib/puppet/concat/_etc_supervisord.conf/fragments/02_supervisord_main]/ensure: defined content as '{md5}ee667fb8f3cea08d885e29755cccce32'
==> default: Info: /Stage[main]/Supervisord::Config/Concat::Fragment[supervisord_main]/File[/var/lib/puppet/concat/_etc_supervisord.conf/fragments/02_supervisord_main]: Scheduling refresh of Exec[concat_/etc/supervisord.conf]
==> default: Notice: /Stage[main]/Supervisord::Config/Concat[/etc/supervisord.conf]/File[/var/lib/puppet/concat/_etc_supervisord.conf/fragments.concat.out]/ensure: created
==> default: Notice: /Stage[main]/Supervisord::Config/Concat[/etc/supervisord.conf]/File[/var/lib/puppet/concat/_etc_supervisord.conf/fragments.concat]/ensure: created
==> default: Notice: /Stage[main]/Supervisord::Config/Concat::Fragment[supervisord_unix]/File[/var/lib/puppet/concat/_etc_supervisord.conf/fragments/01_supervisord_unix]/ensure: defined content as '{md5}2a371bf6742698a806a2ab9879d60540'
==> default: Info: /Stage[main]/Supervisord::Config/Concat::Fragment[supervisord_unix]/File[/var/lib/puppet/concat/_etc_supervisord.conf/fragments/01_supervisord_unix]: Scheduling refresh of Exec[concat_/etc/supervisord.conf]
==> default: Notice: /Stage[main]/Supervisord::Config/Concat[/etc/supervisord.conf]/Exec[concat_/etc/supervisord.conf]/returns: executed successfully
==> default: Notice: /Stage[main]/Supervisord::Config/Concat[/etc/supervisord.conf]/Exec[concat_/etc/supervisord.conf]: Triggered 'refresh' from 4 events
==> default: Notice: /Stage[main]/Supervisord::Config/Concat[/etc/supervisord.conf]/File[/etc/supervisord.conf]/ensure: defined content as '{md5}b112084fc510d84abb5b378d21e59b4b'
==> default: Info: Class[Supervisord::Config]: Scheduling refresh of Class[Supervisord::Service]
==> default: Info: Class[Supervisord::Service]: Scheduling refresh of Service[supervisord]
==> default: Error: Could not start Service[supervisord]: Execution of '/etc/init.d/supervisord start' returned 1: 
==> default: Error: /Stage[main]/Supervisord::Service/Service[supervisord]/ensure: change from stopped to running failed: Could not start Service[supervisord]: Execution of '/etc/init.d/supervisord start' returned 1: 
==> default: Notice: /Stage[main]/Supervisord::Service/Service[supervisord]: Triggered 'refresh' from 1 events
==> default: Notice: /Stage[main]/Supervisord::Reload/Exec[supervisorctl_reread]: Dependency Service[supervisord] has failures: true
==> default: Notice: /Stage[main]/Supervisord::Reload/Exec[supervisorctl_update]: Dependency Service[supervisord] has failures: true
==> default: Warning: /Stage[main]/Supervisord::Reload/Exec[supervisorctl_reread]: Skipping because of failed dependencies
==> default: Warning: /Stage[main]/Supervisord::Reload/Exec[supervisorctl_update]: Skipping because of failed dependencies

supervisord isn't in the config.yaml, it is however used in the manifest. checking supervisord's git repo, they did a commit last night (12 hours ago) and even though it was to fix something, I can't help but think it's broken something else where.

**Comment by fain182** [comment]
I confirm this problem.

**Comment by liuggio** [comment]
d'oh we are on stuck for this.

**Comment by alxndrmlr** [comment]
Likewise. I'm trying to dig into it right now.

**Comment by MPParsley** [comment]
Same issue here.

**Comment by alxndrmlr** [comment]
```
[03:37 PM]-[vagrant@packer-virtualbox-iso-1422596416]-[~] 
$ pip list
apt-xapian-index (0.44)
chardet (2.0.1)
command-not-found (0.2.44)
Warning: cannot find svn location for distribute===0.6.24dev-r0
distribute (0.6.24dev-r0)
GnuPGInterface (0.3.2)
language-selector (0.1)
meld3 (1.0.2)
pip (6.0.8)
pycurl (7.19.0)
python-apt (0.8.3ubuntu7.2)
python-debian (0.1.21ubuntu1)
setuptools (0.6rc11)
supervisor (3.1.3)
ufw (0.31.1-1)
unattended-upgrades (0.1)

[03:37 PM]-[vagrant@packer-virtualbox-iso-1422596416]-[~] 
$ sudo /etc/init.d/supervisord start
Starting supervisor: Traceback (most recent call last):
  File "/usr/local/bin/supervisord", line 5, in <module>
    from pkg_resources import load_entry_point
  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 2707, in <module>
    working_set.require(__requires__)
  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 686, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 584, in resolve
    raise DistributionNotFound(req)
pkg_resources.DistributionNotFound: meld3>=0.6.5

[03:37 PM]-[vagrant@packer-virtualbox-iso-1422596416]-[~] 
```

If go in and upgrade ```distribute (0.6.24dev-r0)``` to ```distribute (0.7.3)``` then I can start the supervisor.

```
[03:40 PM]-[vagrant@packer-virtualbox-iso-1422596416]-[~] 
$ sudo pip install distribute --upgrade
The directory '/home/vagrant/.cache/pip/log' or its parent directory is not owned by the current user and the debug log has been disabled. Please check the permissions and owner of that directory. If executing pip with sudo, you may want the -H flag.
The directory '/home/vagrant/.cache/pip/http' or its parent directory is not owned by the current user and the cache has been disabled. Please check the permissions and owner of that directory. If executing pip with sudo, you may want the -H flag.
The directory '/home/vagrant/.cache/pip/http' or its parent directory is not owned by the current user and the cache has been disabled. Please check the permissions and owner of that directory. If executing pip with sudo, you may want the -H flag.
Collecting distribute from https://pypi.python.org/packages/source/d/distribute/distribute-0.7.3.zip#md5=c6c59594a7b180af57af8a0cc0cf5b4a
  Downloading distribute-0.7.3.zip (145kB)
    100% |################################| 147kB 494kB/s 
^[[ACollecting setuptools>=0.7 (from distribute)
  Downloading setuptools-14.3.1-py2.py3-none-any.whl (501kB)
    100% |################################| 503kB 575kB/s 
Installing collected packages: setuptools, distribute
  Found existing installation: setuptools 0.6rc11
    DEPRECATION: Uninstalling a distutils installed project (setuptools) has been deprecated and will be removed in a future version. This is due to the fact that uninstalling a distutils project will only partially uninstall the project.
    Uninstalling setuptools-0.6rc11:
      Successfully uninstalled setuptools-0.6rc11

  Found existing installation: distribute 0.6.24dev-r0
    Uninstalling distribute-0.6.24dev-r0:
      Successfully uninstalled distribute-0.6.24dev-r0
  Running setup.py install for distribute
Successfully installed distribute-0.7.3 setuptools-14.3.1

[03:40 PM]-[vagrant@packer-virtualbox-iso-1422596416]-[~] 
$ sudo /etc/init.d/supervisord start
Starting supervisor: supervisord.
```

Looks like the issue is that pip is install a dev branch of distribute.

Thought I'd share, still trying to look into where / why / how its pulling the dev version in. (I don't really know puppet that well just trying to dig into ti)

**Comment by alxndrmlr** [comment]
Ah okay so distribute isn't the culprit.

https://pythonhosted.org/distribute/_sources/index.txt

```
Welcome to Distribute's documentation!
======================================

`Distribute` is a deprecated fork of the `Setuptools` project.

Since the Setuptools 0.7 release, Setuptools and Distribute have merged
and Distribute is no longer being maintained. All ongoing effort should
reference the
`Setuptools project <https://bitbucket.org/pypa/setuptools>`_ and the
`Setuptools documentation <https://pythonhosted.org/setuptools>`_.
```

Looking at setuptools now

**Comment by jtreminio** [comment]
Well this is going to screw up so many people's VMs.

**Comment by alxndrmlr** [comment]
I've hacked together a fix to unblock me for now.

I've modified ```puphpet/manifests/python/pip.pp``` to contain the following

```
class puphpet::python::pip {

  Exec { path => [ '/usr/bin/', '/usr/local/bin', '/bin', '/usr/local/sbin', '/usr/sbin', '/sbin' ] }

  #  if ! defined(Package['python-setuptools']) {
  #    package { 'python-setuptools': }
  #  }
  #
  #  exec { 'easy_install pip':
  #    unless  => 'which pip',
  #    require => Package['python-setuptools'],
  #  }

  # Hack to get the latest version of setuptool
  # See https://github.com/puphpet/puphpet/issues/1492
  exec { 'install_setuptools':
    command => "curl https://bootstrap.pypa.io/ez_setup.py | sudo python",
    cwd     => '/tmp',
    before  => Exec['easy_install pip']
  }

  exec { 'easy_install pip':
    unless  => 'which pip',
  }

  if $::osfamily == 'RedHat' {
    exec { 'rhel pip_provider_name_fix':
      command     => 'alternatives --install /usr/bin/pip-python pip-python /usr/bin/pip 1',
      subscribe   => Exec['easy_install pip'],
      unless      => 'which pip-python',
    }
  }

}
```

@jtreminio I realize this probably not the best way to solve the issue, but I thought I'd share it for the others that are running into it currently so as to unblock them in the interim.

**Comment by Zeromax** [comment]
run into this as well... Let me know how to help.

**Comment by Zeromax** [comment]
@alxndrmlr thanks for this fix. Now I can get it up again ;)

**Comment by alxndrmlr** [comment]
No problem @Zeromax!

**Comment by rschaufler** [comment]
thank you @alxndrmlr ! Most of our VM's broke due to this bug and your patch was a real life saver

**Comment by webermax** [comment]
Same here - any workaround available?

**Comment by alxndrmlr** [comment]
@webermax does the fix / workaround I added above not work for you?

**Comment by epastrana** [comment]
Thanks @alxndrmir. That fixed the erro. I can start working again. :)

**Comment by vikmind** [comment]
I resolved same problem with solution from https://github.com/puphpet/puphpet/issues/1506 proposed by @adamlundrigan: only install ``python-meld3`` package

**Comment by sanjeevdharap** [comment]
Going back to meld3 v1.0.1 (pip install 'meld3 == 1.0.1') in the vagrant vm worked for me. supervisorctl no longer gives me an error.

**Comment by mrconnerton** [comment]
Adding python-meld3 as package to install resolved for me as well.

**Comment by paolooo** [comment]
Unfortunately, installing `python-meld3` didn't worked for me :(.

```
# config.yml
...
python:
    install: '1'
    packages:
      - meld3
    versions: {  }
...
```

```
[vagrant@localhost]-[~] $ pip list
/usr/lib/python2.6/site-packages/pip-6.1.0-py2.6.egg/pip/_vendor/requests/packages/urllib3/util/ssl_.py:79: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. For more information
, see https://urllib3.readthedocs.org/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
You are using pip version 6.1.0, however version 6.1.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
distribute (0.6.10)
ethtool (0.6)
iniparse (0.3.1)
iwlib (1.0)
meld3 (1.0.2)
pip (6.1.0)
pycurl (7.19.0)
pygpgme (0.1)
setuptools (0.6rc11)
supervisor (3.1.3)
urlgrabber (3.9.1)
yum-metadata-parser (1.1.2)
yum-presto (0.4.4)
```

**Comment by paolooo** [comment]
Disregard my previous comment. @alxndrmlr's solution works for me, but I removed the `sudo` in `command => "curl https://bootstrap.pypa.io/ez_setup.py | sudo python",` line, because it throws an error.

```
==> default: Notice: /Stage[main]/Puphpet::Python::Pip/Exec[install_setuptools]/returns: sudo: sorry, you must have a tty to run sudo
==> default: Notice: /Stage[main]/Puphpet::Python::Pip/Exec[install_setuptools]/returns:   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
==> default: Notice: /Stage[main]/Puphpet::Python::Pip/Exec[install_setuptools]/returns:                                  Dload  Upload   Total   Spent    Left  Speed
```

So, my version is:

```
# puphpet/puppet/modules/puphpet/manifests/python/pip.pp 

class puphpet::python::pip {

  Exec { path => [ '/usr/bin/', '/usr/local/bin', '/bin', '/usr/local/sbin', '/usr/sbin', '/sbin' ] }

  #  if ! defined(Package['python-setuptools']) {
  #    package { 'python-setuptools': }
  #  }
  #
  #  exec { 'easy_install pip':
  #    unless  => 'which pip',
  #    require => Package['python-setuptools'],
  #  }

  # Hack to get the latest version of setuptool
  # See https://github.com/puphpet/puphpet/issues/1492
  exec { 'install_setuptools':
    command => "curl https://bootstrap.pypa.io/ez_setup.py | python",
    cwd     => '/tmp',
    before  => Exec['easy_install pip']
  }

  exec { 'easy_install pip':
    unless  => 'which pip',
  }

  if $::osfamily == 'RedHat' {
    exec { 'rhel pip_provider_name_fix':
      command     => 'alternatives --install /usr/bin/pip-python pip-python /usr/bin/pip 1',
      subscribe   => Exec['easy_install pip'],
      unless      => 'which pip-python',
    }
  }

}
```

Thanks @alxndrmlr :)

**Comment by SidFerreira** [comment]
@paolooo Your solution worked for me too. Thanks to @alxndrmlr too

**Comment by karlingen** [comment]
+1

**Comment by MPParsley** [comment]
Applied the latest patch by @paolooo. Installation now runs fine.

**Comment by paolooo** [comment]
Thanks @SidFerreira, @karlingen,  and @MPParsley but all credits goes to @alxndrmlr. Thanks to @alxndrmlr :)

**Comment by dragosprotung** [comment]
A colleague had the same issue and we discovered that they cloned the repository with autcrlf on under Windows. After fixing the line endings everything worked fine

**Comment by franz-josef-kaiser** [comment]
OS X 10.10 - same problem. Had to manually start `sudo service php-fpm start` that service.

**Comment by cnphyt** [comment]
+1

**Comment by pbylina** [comment]
```
[root@localhost vagrant]# supervisorctl
Traceback (most recent call last):
  File "/usr/bin/supervisorctl", line 5, in <module>
    from pkg_resources import load_entry_point
  File "/usr/lib/python2.6/site-packages/pkg_resources.py", line 2655, in <module>
    working_set.require(__requires__)
  File "/usr/lib/python2.6/site-packages/pkg_resources.py", line 648, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/usr/lib/python2.6/site-packages/pkg_resources.py", line 546, in resolve
    raise DistributionNotFound(req)
pkg_resources.DistributionNotFound: meld3>=0.6.5
[root@localhost vagrant]# pip freeze | grep meld3
meld3==1.0.2
[root@localhost vagrant]# pip install meld3==1.0.0
Collecting meld3==1.0.0
  Using cached meld3-1.0.0.tar.gz
Installing collected packages: meld3
  Found existing installation: meld3 1.0.2
    Uninstalling meld3-1.0.2:
      Successfully uninstalled meld3-1.0.2
  Running setup.py install for meld3
Successfully installed meld3-1.0.0
[root@localhost vagrant]# supervisorctl
http://localhost:9001 refused connection
supervisor>
```

After downgrading meld3 to 1.0.0 supervisord works.

**Comment by ReservedDeveloper** [comment]
This all seems to be tied to the warning re: InsecurePlatformWarning from urllib3. See [here](https://urllib3.readthedocs.org/en/latest/security.html#insecureplatformwarning).

I noticed by default this system uses Python 2.7.6. Unfortunately, that seems to be the default version in Ubuntu 14.04. Using the [PPA from Deadsnake](https://launchpad.net/~fkrull/+archive/ubuntu/deadsnakes-python2.7) to install 2.7.10, and managing any pips via an exec-* file seems to have resolved the error message. YMMV.

**Comment by therobyouknow** [comment]
Thanks for the solutions, some of them mentioned here and linked to from another issue worked for my situation.

I upgraded to MacOS X El Capitan 10.11 and wanted to test that my VM-based development environment still worked. I found that the usual vagrant up was fine but rebuilding a VM from scratch failed with similar errors to those mentioned here, e.g.: "/Stage[main]/Supervisord::Reload/Exec[supervisorctl_update]: supervisorctl update returned 1 instead of one of [0,2]"

I found the following to work:
Using the custom code for # puphpet/puppet/modules/puphpet/manifests/python/pip.pp - here: https://github.com/puphpet/puphpet/issues/1492#issuecomment-90869454
AND putting the following at the end of my config.yaml
```
python:
    install: '1'
    packages:
      - meld3
    versions: {  }
```

My setup:
MacOS X El Capitan 10.11
MacBookPro 13" Retina Early 2013. 8Gb RAM
Vagrant 1.7.2
VirtualBox 4.3.28 r100309
puphpet manifest from 2015 06 03 (I think)

I think that this issue might have also been there before upgrading to El Capitan but can't be sure as hadn't tested rebuilding recently, i.e. just before upgrading to El Capitan. It seems that at the time, a puphpet manifest works but later on that same manifest can fail when used to do a rebuild and this is because the manifest has external dependencies for things on line which can change themselves. Fortunately it seems like these can be overcome with solutions in issues like this one.

Thank you very much for your help.

**Comment by vikas027** [comment]
I had to run this command to overcome this error `pip install meld3`

**Comment by justin-hackin** [comment]
pip install pyopenssl ndg-httpsclient pyasn1 worked for me thanks @ReservedDeveloper

**Comment by sabrio** [comment]
For me seems to be the python version: 2.7.*. I changed it to 2.6 it works `alternatives --set python /usr/bin/python2.6`
